src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,156 @@
+				ExtractTarGzipFile( downloadPath, extractFolder );
+			} else {
+				throw new Exception( "Unknown archive type" );
+			}
+		} catch( Exception ex ) {
+			Directory.Delete( extractFolder, true );
+			throw new BmxException( "Failed to update with new files", ex );
+		} finally {
+			File.Delete( downloadPath );
+		}
+
+		try {
+			File.Move( currentFilePath, backupPath );
+		} catch( IOException ex ) {
+			Directory.Delete( extractFolder, true );
+			throw new BmxException( "Could not remove the old version. Please try again with elevated permissions.", ex );
+		} catch {
+			Directory.Delete( extractFolder, true );
+			throw new BmxException( "BMX could not update" );
+		}
+
+		try {
+			foreach( string file in Directory.GetFiles( extractFolder ) ) {
+				string destinationFile = Path.Combine( currentDirectory, Path.GetFileName( file ) );
+				File.Move( file, destinationFile );
+			}
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentFilePath );
+			throw new BmxException( "BMX could not update with the new version", ex );
+		} finally {
+			Directory.Delete( extractFolder, true );
+		}
+	}
+
+	private static string GetOSFileName() {
+
+		if( RuntimeInformation.IsOSPlatform( OSPlatform.Windows ) ) {
+			return "bmx-win-x64.zip";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.OSX ) ) {
+			return "bmx-osx-x64.tar.gz";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.Linux ) ) {
+			return "bmx-linux-x64.tar.gz";
+		} else {
+			throw new BmxException( "New version does not support you current OS" );
+		}
+	}
+
+	public static void Cleanup() {
+		if( Directory.Exists( BmxPaths.OLD_BMX_VERSIONS_PATH ) ) {
+			try {
+				Directory.Delete( BmxPaths.OLD_BMX_VERSIONS_PATH, true );
+			} catch( Exception ) {
+				Console.Error.WriteLine( "WARNING: Failed to delete old version files" );
+			}
+		}
+	}
+
+	private static void ExtractTarGzipFile( string compressedFilePath, string decompressedFilePath ) {
+		string tarPath = Path.Combine( decompressedFilePath, "bmx.tar" );
+		using( FileStream compressedFileStream = File.Open(
+			compressedFilePath,
+			FileMode.Open,
+			FileAccess.Read,
+			FileShare.Read )
+		) {
+			using FileStream outputFileStream = File.Create( tarPath );
+			using var decompressor = new GZipStream( compressedFileStream, CompressionMode.Decompress );
+			decompressor.CopyTo( outputFileStream );
+		}
+
+		try {
+			TarFile.ExtractToDirectory( tarPath, decompressedFilePath, true );
+		} finally {
+			File.Delete( tarPath );
+		}
+	}
+
+	private static void ExtractZipFile( string compressedFilePath, string decompressedFilePath ) {
+		using ZipArchive archive = ZipFile.OpenRead( compressedFilePath );
+		foreach( ZipArchiveEntry entry in archive.Entries ) {
+			string destinationPath = Path.GetFullPath( Path.Combine( decompressedFilePath!, entry.FullName ) );
+			if( destinationPath.StartsWith( decompressedFilePath!, StringComparison.Ordinal ) ) {
```

Comment:
`decompressedFilePath` isn't nullable
```suggestion
			string destinationPath = Path.GetFullPath( Path.Combine( decompressedFilePath, entry.FullName ) );
			if( destinationPath.StartsWith( decompressedFilePath, StringComparison.Ordinal ) ) {
```